packages/devtools/lib/src/message_manager.dart

```diff
@@ -0,0 +1,136 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+
+import 'globals.dart';
+import 'ui/elements.dart';
+import 'ui/primer.dart';
+
+class MessageManager {
+  MessageManager();
+
+  static const _generalId = 'general';
+
+  final _container = CoreElement.from(queryId('messages-container'));
+
+  /// Maps screen ids to their respective messages.
+  ///
+  /// Messages that do not pertain to a specific screen will be stored under the
+  /// key [_generalId].
+  final Map<String, Set<Message>> _messages = {};
+
+  final List<String> _dismissedMessageIds = [];
+
+  void showMessagesForScreen(String screenId) {
+    _messages[screenId]?.forEach(_showMessage);
+  }
+
+  void _showMessage(Message message) {
+    if (_dismissedMessageIds.contains(message.id)) return;
+    _container.add(message.flash);
+  }
+
+  void removeAll() {
+    _container.clear();
+    // Remove all error messages.
+    _messages[_generalId]
+        ?.removeWhere((m) => m.messageType == MessageType.error);
+  }
+
+  void addMessage(Message message, {String screenId = _generalId}) {
+    message.onDismiss.listen((_message) {
+      if (_message.id != null) {
+        _dismissedMessageIds.add(_message.id);
+      }
+      _messages[screenId]?.remove(_message);
+    });
+
+    // ignore: prefer_collection_literals
+    _messages[screenId] ??= Set()..add(message);
```

Comment:
Is this correct? I'm having trouble parsing it. Is this equivalent to:

```
_messages.putIfAbsent(screenId, () => Set()).add(message);
```

Reply:
whoops I am missing parens around `_messages[screenId] ??= Set()`. If those had been there, then yes, they would have been equivalent. changing to use `putIfAbsent` instead.